Add ILLink targets and tests

sbomer opened the pull request:
This adds linker targets, mostly adapted from https://github.com/mono/linker/blob/9766e2b2406a3d44cbf05e49ab2bd19b535e3e5a/src/ILLink.Tasks/ILLink.Tasks.targets, to run the linker after `ComputeFilesToPublish`.

A few things still need to change (but I wanted to get what I have so far out for review):
- ~~Change property names to agreed-upon strings~~
- ~~Update deps file generation in response to https://github.com/dotnet/sdk/pull/3086. I think this needs https://github.com/dotnet/sdk/pull/3113 or another update PR to be merged.~~
- ~~Update the linker:~~
  - ~~to include `UsingTask` for other tasks in the assembly (namely, `ComputeManagedAssemblies`)~~
  - ~~for `ReferencePath` support~~
  - ~~with a net472 build of the task dll~~

~~When I run these tests locally with a recent version of the linker, the checks pass up to the deps file asserts.~~
All checks are passing.

@nguerrera, @fadimounir, @swaroop-sridhar PTAL

sbomer committed:
Add ILLink targets and tests

swaroop-sridhar commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -175,6 +176,17 @@ Copyright (c) .NET Foundation. All rights reserved.
+    -->
+  <Target Name="ILTransform"
+          Condition="'$(_TargetFrameworkVersionWithoutV)' >= '3.0' And '$(TargetFrameworkIdentifier)' == '.NETCoreApp'"
+          DependsOnTargets="$(ILTransformDependsOn)" />
```
Why is this only for `.NetCoreApp`? How about, if there are whole unused IL files that are not reachable from any public entrypoints in a `.netstandard` dll?  (ex: partial classes) ?

sbomer replied:
I think you're right that it would make sense for netstandard, and the linker behavior is probably similar to the portable app case.

sbomer replied:
I think that since linking happens during publish, the normal workflow of packaging a netstandard library would not run the linker, and it would take some work by library authors to package a linked library. I agree that it makes sense not to block it here.

swaroop-sridhar commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -0,0 +1,110 @@
+          Condition=" '$(LinkDuringPublish)' == 'true' "
+          DependsOnTargets="RunILLink">
+
+    <!-- For now, use ResolvedFileToPublish as input/output. This
```
The linker works with `ResolvedFileToPublish` as input, and I think that's the way it should be for now.

But I'm curious why crossgen works differently. Crossgen targets run after `_ComputeResolvedFilesToPublishTypes` target, and processes/updates the files bifurcated into `_ResolvedFileToPublishPreserveNewest`  and `_ResolvedFileToPublishAlways` items. However, the processing /input-outputs for the two lists there are similar. So, why doesn't crossgen similarly process `ResolvedFileToPublish` ?

sbomer replied:
Good question - @fadimounir can you answer this?

fadimounir replied:
When I started implementing this, i was under the impression that I would need to add some logic to the crossgen task to distinguish between the 2 lists, and only compile things in the PreserveNewest if they were out of date.
Things have evolved since then, and thinking about it now, I believe we should be able to just make crossgen work with the 'ResolvedFileToPublish', before the split that '_ComputeResolvedFilesToPublishTypes' does.
I'll make that change. For the linker, i think this looks correct.
The linker does one operation with multiple input files. Crossgen does multiple operations, each with its own input file, so splitting the 'ResolvedFilesToPublish' by type doesn't make sense here i believe

swaroop-sridhar commented:
Looks good.

sbomer committed:
Enable and test deps file trimming

Deps file trimming doesn't work on ProjectReferences, so this changes
the referenced test project to be a PackageReference.

We also disable use of the build dependency file when linking, as the
publish dependency file needs to be generated based on what the linker
decided to remove.

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -0,0 +1,115 @@
+<!--
```
Not sure bout this, but i think this file should be added the the ILLink nuget package as a Sdk.targets file. I see other SDKs do that too. @nguerrera What are your thoughts here?

fadimounir replied:
Example: see Microsoft.NET.Sdk.Razor, or Microsoft.NET.Sdk.Web

wli3 replied:
i prefer to have a separate package. Although in this PR, ILTransform is part of Publish, that means it is not optional, does it have to?

sbomer replied:
I'd like it to be in the SDK if possible. With the out-of-band linker package, we shipped our own targets that we had trouble keeping in sync with the SDK (see https://github.com/dotnet/sdk/issues/2583 for a taste of the issues).
- One advantage that you already pointed out is that we don't have to use BeforeTargets/AfterTargets as an extension point.
- There's also the dependency in `_ComputeUseBuildDependencyFile`: https://github.com/dotnet/sdk/pull/3125/files/3f83d196a8476f3889e2191e00c465e0c113b4a6#diff-7dc833d02f198e5a9f0ab88e229db18aR781.
- Finally, the use of `ResolvedFileToPublish` as an input is something that I hope we can replace with a more well-defined list of runtime assemblies (see https://github.com/dotnet/sdk/issues/3109), which is another case where we have a pretty tight coupling with the SDK.

sbomer replied:
The mosnt important thing is that we have tests for the logic in these targets in the SDK repo. As long as we do that, it could live in either place but I think here makes sense. Curious what you think @nguerrera.

nguerrera replied:
Given that there is mutation of state, I would prefer to keep this here so we can have a pure contract with the linker.

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -0,0 +1,115 @@
+    -->
+  <Target Name="ILLink"
+          Condition=" '$(LinkDuringPublish)' == 'true' "
+          DependsOnTargets="RunILLink">
```
>RunILLink [](start = 28, length = 9)

The convention i see in the main Microsoft.NET.Publish.targets is to use underscores as name prefixes for "sub-targets" executed from a main target. We should probably do the same here (_ILLink, _RunILLink, ...)

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -0,0 +1,115 @@
+           Inputs="@(_ManagedAssembliesToLink);@(LinkerRootDescriptors);@(ReferencePath)"
+           Outputs="$(_LinkSemaphore)">
+
+     <Delete Files="@(_LinkedResolvedFileToPublishCandidates)" />
```
><Delete Files="@(_LinkedResolvedFileToPublishCandidates)" /> [](start = 5, length = 60)

I thought we were going to RemoveDir, no?

sbomer replied:
I believe @nguerrera preferred to delete individual files rather than directories to be conservative and reduce risk of deleting things we shouldn't be. Please correct me if I'm wrong. :)

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -0,0 +1,115 @@
+   <Target Name="RunILLink"
+           DependsOnTargets="_ComputeManagedAssembliesToLink"
+           Inputs="@(_ManagedAssembliesToLink);@(LinkerRootDescriptors);@(ReferencePath)"
+           Outputs="$(_LinkSemaphore)">
```
What if the set of inputs are unchanged here, but the user decides to use a different more aggressive linking mode by changing some property in the csproj. Wouldn't msbuild skip this target here, given that all inputs/outputs are unchanged?

sbomer replied:
Good question - I see some targets that have `$(MSBuildAllProjects)` as an input. I believe that would have the desired effect. @nguerrera is that the correct approach? Is there a way to accomplish the same thing for properties passed on the command-line?

rainersigwald replied:
It all depends on how careful you want to be and whether you expect scenarios to be common or not.

`$(MSBuildAllProjects)` catches changes to properties _specified in project files_. That's most projects, but it doesn't include things set on the command line. Here, that'd be something like `dotnet build -p:ExtraLinkerArgs={whatever}`.

That can also be handled, if you think it's likely to be an important scenario. Right now, it's done only for CoreCompile, added in Microsoft/msbuild#3978.

You know the scenario better, but I lean toward including `$(MSBuildAllProjects)` and leaving it at that until/unless there are significant user complaints.

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -0,0 +1,115 @@
+      <ResourceCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <RuntimeCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <RuntimeTargetsCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <RuntimePackAsset Remove="@(_RemovedManagedAssemblies)" />
```
This is the kind of change that would be undesirable if we moved this targets file to the ILLink nuget package, because if the names here change, this would be breaking.
@nguerrera, how can we ensure that name changes here won't break other tools/SDKs that depend on this? CI tests good enough?

wli3 replied:
if other sdks to depend on it is expected, it should be public and follow backward compact rules

nguerrera replied:
This is exactly why we are putting the targets portion in the SDK. The contract to the linker is pure, no global mutations. Yes, other tools might do this sort of thing, but we specifically cut defining a strict contract for such tools. The linker will be first class and any mutation of the build state will happen in dotnet/sdk. This + sufficient tests will ensure that we don't just break the linker as we used to when it was it's own nuget package. I want to make it possible to write these sort of things outside of the sdk, but it is a hard problem and specifically not one we are going to solve as part of this work.

fadimounir commented:
Looks good overall, modulo the question regarding where the targets file should be, and any related side-effect.

sbomer committed:
Support old deps generation logic

sbomer committed:
Update property names and respond to PR feedback

The current public-facing linker properties are now:
- PublishTrimmed
- TrimmerRootDescriptors

The IL transform extension point has been named ILTransformForPublish,
to make it clear that it only runs on publish.

This also enables incremental build for project file changes, and lets
the linker run on projects targeting .NETStandard.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -175,6 +176,17 @@ Copyright (c) .NET Foundation. All rights reserved.
 
+  <!--
+    ============================================================
+                                        ILTransformForPublish
```
I'd rather not advertise an extension point, and treat that as a separate feature.

It feels weird that this is tied to the TFM as well. Also, publishing for .NETSTandard is very weird (I wish it didn't work), and there is no .NETStandard 3.0, the current version aligned with 3.0 timeframe is 2.1. I think we should limit to NETCoreApp 3.0+ alone. But I think this limit should be applied directly to the linker and not to some generic extensibility point.

sbomer replied:
I removed this, and made `_ILLink` a direct dependency of `ComputeAndCopyFilesToPublishDirectory`.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -0,0 +1,121 @@
+             RootAssemblyNames="@(IntermediateAssembly->'%(Filename)')"
+             RootDescriptorFiles="@(TrimmerRootDescriptors)"
+             OutputDirectory="$(IntermediateLinkDir)"
+             ExtraArgs="$(ExtraLinkerArgs) --skip-unresolved true" />
```
Should this be ExtraTrimmerArgs if the other public facing options are "Trimmer"?

sbomer replied:
Named it `_ExtraTrimmerArgs`. I don't want to advertise this as a public facing option, and am hoping we can stop using it in the future in favor of defaults that do the right thing. Taking a property here makes it easier to test and makes it easier for us to give users workarounds in case they encounter issues.

sbomer committed:
Respond to PR feedback

- Ensure trailing slash in linker output directory
- Disable old deps file generation logic
- Don't support netstandard
- ExtraLinkerArgs -> _ExtraTrimmerArgs
- Remove ILTransformForPublish extension point

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -15,9 +15,13 @@ Copyright (c) .NET Foundation. All rights reserved.
 
   <PropertyGroup>
     <ILTransformForPublishDependsOn>$(ILTransformForPublishDependsOn);_ILLink</ILTransformForPublishDependsOn>
```
Is this dead now?

sbomer replied:
Yes, thanks!

sbomer committed:
Remove left-over ILTransformForPublishDependsOn property

sbomer commented:
@nguerrera I've responded to all of your feedback. Does it look good to merge? I'd like to get it in so that I can make follow-up changes independently.

sbomer commented:
Looks like the SDK uptake issue has been unblocked: https://github.com/dotnet/sdk/pull/3153. Removing the workarounds from this PR.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -0,0 +1,120 @@
+    <_LinkSemaphore>$(IntermediateOutputPath)Link.semaphore</_LinkSemaphore>
+    <!-- Disable old deps file generation logic until
+         https://github.com/dotnet/sdk/issues/3098 is fixed. -->
+    <DepsFileGenerationMode>new</DepsFileGenerationMode>
```
I don't think you need this anymore if we can merge #3152 first

sbomer replied:
@nguerrera mind if I do that in a follow-up PR? I'd like to get this change in sooner rather than later.

nguerrera replied:
Sure

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:
```diff
@@ -0,0 +1,120 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <TrimmerRootDescriptors Include="$(TrimmerRootDescriptors)" />
```
Are these files? It is weird to have a "public" item and property of the same name. Is there a reason why the user wouldn't provide these directly as items?

Also, item names are generally singular (though it is not 100% consistent) so this is more of a nit.

sbomer replied:
They are files. In the OOB package we supported this to make it easy to test on the command-line. I'll get rid of this in a follow-up and we can iterate there.

sbomer commented:
@nguerrera I know you're very busy atm - thank you for finding time to review this!

sbomer commented:
@nguerrera please merge when you get a chance!
